### PR TITLE
[py-scikit-optimize] added version 0.9.0

### PR DESCRIPTION
**Pull request by qwertos**

### CODE REVIEW

**Review by liuyangzhuan — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-scikit-optimize/package.py [hunk 1]
```diff
@@ -21,17 +21,23 @@ class PyScikitOptimize(PythonPackage):
...
     version("0.5.2", sha256="1d7657a4b8ef9aa6d81e49b369c677c584e83269f11710557741d3b3f8fa0a75")
 
     variant("plots", default=True, description="Build with plot support from py-matplotlib")
-    variant("gptune", default=False, description="Build with patches for GPTune")
+    variant("gptune", default=False, description="Build with patches for GPTune", when="@0.5.2")
 
     depends_on("py-setuptools", type="build")
     depends_on("py-numpy", type=("build", "run"))
-    depends_on("py-scipy@0.14.0:", type=("build", "run"))
+    depends_on("py-numpy@1.12:", when="@0.9:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scikit-optimize/package.py]:
```suggestion
    depends_on("py-numpy@1.13.3:", when="@0.9:", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-scikit-optimize/package.py [hunk 2]
```diff
@@ -21,17 +21,23 @@ class PyScikitOptimize(PythonPackage):
...
+    depends_on("py-scipy@0.14:", type=("build", "run"))
+    depends_on("py-scipy@0.18:", when="@0.9:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scikit-optimize/package.py]:
```suggestion
    depends_on("py-scipy@0.19.1:", when="@0.9:", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-scikit-optimize/package.py [hunk 3]
```diff
@@ -21,17 +21,23 @@ class PyScikitOptimize(PythonPackage):
...
     depends_on("py-scikit-learn@0.19.1:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scikit-optimize/package.py:33]: 0.20+ now

**Review by adamjstewart — CHANGES_REQUESTED** [review]
See `setup.py` for the new min versions.

If you want to sort deps in the same order as in setup.py and from newest to oldest it makes reviewing slightly easier.

**Comment by qwertos** [comment]
@adamjstewart The versions are different between `setup.py` and `pyproject.toml` what is authoritative when there is a difference?

**Comment by adamjstewart** [comment]
In this particular case, they are both authoritative. The requirements in `pyproject.toml` are the versions that need to be present when building the package. And the requirements in `setup.py` are the versions that need to be present at run-time. In the future I'm hoping Spack can handle different requirements for build vs. run-time like pip can (@alalazo) but for now you should take the stricter of the two requirements if they differ.

**Review by adamjstewart — DISMISSED** [review]

File: var/spack/repos/builtin/packages/py-scikit-optimize/package.py [hunk 4]
```diff
@@ -21,17 +21,24 @@ class PyScikitOptimize(PythonPackage):
     maintainers("liuyangzhuan")
 
     version("master", branch="master")
+    version("0.9.0", sha256="77d8c9e64947fc9f5cc05bbc6aed7b8a9907871ae26fe11997fd67be90f26008")
     version("0.5.2", sha256="1d7657a4b8ef9aa6d81e49b369c677c584e83269f11710557741d3b3f8fa0a75")
 
     variant("plots", default=True, description="Build with plot support from py-matplotlib")
-    variant("gptune", default=False, description="Build with patches for GPTune")
+    variant("gptune", default=False, description="Build with patches for GPTune", when="@0.5.2")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scikit-optimize/package.py]:
```suggestion
    variant("gptune", default=False, description="Build with patches for GPTune")
```
Technically this variant shouldn't even be in here, it should be added to the gptune package. The only package that needs this is gptune, which requires the master branch, which is not 0.5.2, hence the failing tests. Let's leave this alone, maybe @liuyangzhuan can move this patch to the gptune package in a separate PR?

Comment (liuyangzhuan) [reply var/spack/repos/builtin/packages/py-scikit-optimize/package.py]: @adamjstewart Could you point out how to add a patch file to the py-scikit-optimize subdirectory from modifying the package.py file of gptune?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-scikit-optimize/package.py]: An example from the `gdal` package:
```python
    depends_on("jasper@1.900.1", patches=[patch("uuid.patch")], when="+jasper")                     
```
The `uuid.patch` file, stored alongside the GDAL `package.py`, will be applied to `jasper` when building it as a dep of `gdal`

Comment (liuyangzhuan) [reply var/spack/repos/builtin/packages/py-scikit-optimize/package.py]: Got it! I will do that in a few days, I'm currently away from my workstation for this week.

Comment (liuyangzhuan) [reply var/spack/repos/builtin/packages/py-scikit-optimize/package.py]: @adamjstewart Can you close this PR and instead take a look at https://github.com/spack/spack/pull/39447 ?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-scikit-optimize/package.py]: Let's keep "adding a new version of py-scikit-optimize" and "moving the patch to a different package" as separate PRs.

File: var/spack/repos/builtin/packages/py-scikit-optimize/package.py [hunk 5]
```diff
@@ -21,17 +21,24 @@ class PyScikitOptimize(PythonPackage):
...
     depends_on("py-setuptools", type="build")
     depends_on("py-numpy", type=("build", "run"))
-    depends_on("py-scipy@0.14.0:", type=("build", "run"))
+    depends_on("py-numpy@1.13.3:", when="@0.9:", type=("build", "run"))
+    depends_on("py-scipy@0.14:", type=("build", "run"))
+    depends_on("py-scipy@0.19.1:", when="@0.9:", type=("build", "run"))
     depends_on("py-scikit-learn@0.19.1:", type=("build", "run"))
-    depends_on("py-pyyaml", when="+gptune", type=("build", "run"))
+    depends_on("py-scikit-learn@0.20:", when="@0.9:", type=("build", "run"))
+    depends_on("py-pyyaml", when="@0.5.2+gptune", type=("build", "run"))
```
Comment (haampie) [inline var/spack/repos/builtin/packages/py-scikit-optimize/package.py]:
```suggestion
```

@adamjstewart what's the point of this line? It's not a dependency of `scikit-optimize`, nor is the `space.patch` patch importing it. It should just be dropped.

**Comment by haampie** [comment]
@qwertos I've taken the liberty to rebase on #39447 and then on `develop`, to push this PR through.

The PRs need to be combined since gptune in fact needs the master branch, which is identical to version 0.9.0.

**Review by haampie — APPROVED** [review]

**Pull request merged**